pkg/deploy/apply.go

```diff
@@ -0,0 +1,476 @@
+	return func(clients *k8sClients, res resourceutil.Resource, deployConfig utils.DeployConfig) error {
+		gvr, err := resourceutil.FromGVKtoGVR(clients.discovery, res.Object.GroupVersionKind())
+		if err != nil {
+			return err
+		}
+
+		// register a watcher and starts to listen for events for the gvr
+		// if res is annotated with awaitCompletionAnnotation
+		var watchEvents <-chan watch.Event
+		startTime := time.Now()
+		awaitCompletionValue, awaitCompletionFound := res.Object.GetAnnotations()[awaitCompletionAnnotation]
+		if awaitCompletionFound {
+			watcher, err := clients.dynamic.Resource(gvr).
+				Namespace(res.Object.GetNamespace()).
+				Watch(context.TODO(), metav1.ListOptions{})
+			if err != nil {
+				return err
+			}
+			watchEvents = watcher.ResultChan()
+			fmt.Printf("Registered a watcher for resource: %s.%s.%s having name %s\n", gvr.Group, gvr.Version, gvr.Resource, res.Object.GetName())
+			defer watcher.Stop()
+		}
+
+		// actually apply the resource
+		if err := apply(clients, res, deployConfig); err != nil {
+			return err
+		}
+
+		// return if no event channel has been set
+		if watchEvents == nil {
+			return nil
+		}
+
+		// parse timeout from annotation value
+		timeout, err := time.ParseDuration(awaitCompletionValue)
+		if err != nil {
+			msg := fmt.Sprintf("Error in %s annotation value: must be a valid duration", awaitCompletionAnnotation)
+			return errors.Wrap(err, msg)
+		}
+
+		// check if res can be handled
+		if _, err := handleResourceCompletionEvent(res, nil, startTime); err != nil {
+			return err
+		}
+
+		// consume watcher events and wait for the resource to complete or exit because of timeout
+		for {
+			select {
+			case event := <-watchEvents:
+				isCompleted, err := handleResourceCompletionEvent(res, &event, startTime)
+				if err != nil {
+					msg := "Error while watching resource events"
+					return errors.Wrap(err, msg)
+				}
+
+				if isCompleted {
+					return nil
+				}
+			case <-time.NewTimer(timeout).C:
+				msg := fmt.Sprintf("Timeout received while waiting for resource %s completion", res.Object.GetName())
+				return errors.New(msg)
+			}
+		}
+	}
+}
+
+// handleResourceCompletionEvent takes the target resource, the watch event and
+// the initial watch time as arguments. It returns (true, nil) when the given
+// resource has completed in the given event. If the event is nil returns (false, nil)
+// when the resource supports events watching otherwise returns (false, error).
+func handleResourceCompletionEvent(res resourceutil.Resource, event *watch.Event, startTime time.Time) (bool, error) {
+	switch res.GroupVersionKind.Kind {
+	case "Job":
+		// only manage watch.Modified events
+		if event == nil || event.Type != watch.Modified {
+			return false, nil
+		}
+		// convert resources into jobs
+		var jobFromRes batchv1.Job
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(res.Object.Object, &jobFromRes); err != nil {
+			return false, err
+		}
```

Comment:
As of the current implementation this conversion cannot be done outside the for loop in `withAwaitableResource` because this function is meant to handle generic resources not jobs only